### PR TITLE
Adds optional DNI field and missing post_code handling in supplier form

### LINKED ISSUE #17268: DNI and Postal code not added when creating a supplier
**Describe the bug**
When creating a supplier, if DNI and Zip/postal code are filled, these fields will not be filled in the created supplier.

**To Reproduce**
1. Go to Supplier, create a new supplier
2. Fill the DNI and Zip/postal code fields and create
3. Open the supplier : DNI and Zip/postal code fields are empty
4. Edit the supplier, add again DNI and Zip code and save
5. Open the supplier : DNI field is still empty, Zip code is filled

>Edit By Khouloud
Other issues reported when selecting Spain country, please see: https://github.com/PrestaShop/PrestaShop/issues/17268#issuecomment-576634443

**Screenshots**
https://drive.google.com/file/d/181rP_-rIIkXvGHGpzEwp2X5l72wzJGsT/view?usp=sharing

**Additional information**
PrestaShop version: 1.7.7.x
PHP version: N/A
Other info : Tested with a shop installed in English, with country "France"

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | Pr modifies Supplier form. Adds missing handling of post_code field. Adds optional DNI field, which in legacy used to be hidden unless country required it (e.g. Spain by default). We make it simply optional for all countries as a temporary fix. The addresses rework in 1.7.8 will possibly update the form to consider country and required fields settings.
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #17268
| How to test?  | Open supplier form and see that there is no DNI field and post code field should be saved correctly.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/17305)

### CODE REVIEW

**Comment by matks** [comment]
@marionf Apparently, we added "DNI" and "postcode" inputs to Supplier page when we migrated it. Probably because it seems to make sense 😄.

It seems it was a bad idea but ... is it ? Can you tell us what you think ?

**Comment by PierreRambaud** [comment]
Supplier has address, it must contain a DNI if needed :sweat_smile:

**Comment by zuk3975** [comment]
Hmm i was migrating suppliers myself. I just added it by mistake, comparing to legacy it doesnt have dni. Ofcourse i can add the field back and implement its handling, its not so hard.

**Comment by marionf** [comment]
@matks the zip code was already there in legacy and make sense as it's an address
However the DNI wasn't there in legacy and not seems logical to have it for a supplier

**Comment by PierreRambaud** [comment]
@marionf the DNI is something mandatory for some countries :thinking:

**Comment by marionf** [comment]
The DNI is the id card ? Not sure a supplier have an id card ?

**Comment by zuk3975** [comment]
> The DNI is the id card ? Not sure a supplier have an id card ?

cmon guys, so, could you tell if it is required or not for sure :smile:  ?

**Comment by PierreRambaud** [comment]
@marionf We need confirmation since DNI can be mandatory each time we have an address form.

**Comment by zuk3975** [comment]
> @marionf We need confirmation since DNI can be mandatory each time we have an address form.

It doesn't depend from location->country required field options nor customer->addresses->requiredFields options (at least according to legacy behaviour)

**Comment by marionf** [comment]
> @marionf the DNI is something mandatory for some countries 

Yes, for a physical person but I don't see the interest of a DNI field for a supplier as it's not a physical person
Unless I am missing something :thinking:

**Comment by matks** [comment]
So ... no DNI ? It's approved ?

**Review by PierreRambaud — DISMISSED** [review]
She said, remove DNI :tada:

**Review by matks — DISMISSED** [review]

**Comment by Robin-Fischer-PS** [comment]
Hi @zuk3975 !

I've an exception when adding a new supplier (call to "getDni" method).
pr17305 getdni error

**Comment by zuk3975** [comment]
> Hi @zuk3975 !
> 
> I've an exception when adding a new supplier (call to "getDni" method).
> pr17305 getdni error

Sorry for such dumb mistake :(, Fixed.

**Review by matks — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by PierreRambaud — DISMISSED** [review]

**Comment by Robin-Fischer-PS** [comment]
Hi @marionf @zuk3975 ! It seems like DNI field was needed for some country, like Spain : It is now impossible to add a supplier address in Spain, because there is an error, I suppose it's because, in address format, DNI in always required in Spain.
The error : 
pr17305 spain dni
The default format address in Spain : 
pr17305 spain format address

**Comment by marionf** [comment]
@zuk3975 

I thought the DNI field wasn't on the legacy page, but it was for some countries, my mistake I am sorry
So, we have to keep the same behavior as legacy page:
Display DNI field for countries where it's mandatory ("Do you need a tax identification number" set to Yes)

**Comment by PierreRambaud** [comment]
@marionf Told you :tada: :tada: :tada: :tada: :joy_cat:

**Review by Progi1984 — DISMISSED** [review]

**Comment by zuk3975** [comment]
@Robin-Fischer-PS,
Reverted the commits. Because its associated with major issue of AddressFormat-Country settigns relation, I did not implement dynamic DNI field rendering, but instead DNI field is always shown as non-required field.

Will create an issue for 1.7.8 to fix addresses and include this page too.
For 1.7.7 Its good to go.

**Comment by zuk3975** [comment]
Related issue: https://github.com/PrestaShop/PrestaShop/issues/17647

**Comment by PierreRambaud** [comment]
The description and the title of this pull request must be updated

**Comment by zuk3975** [comment]
> The description and the title of this pull request must be updated

done

**Review by Progi1984 — DISMISSED** [review]

**Review by jolelievre — DISMISSED** [review]

**Comment by Robin-Fischer-PS** [comment]
Hi ! It's now OK, DNI and Zip code are saved when filled ! QA :heavy_check_mark: 

pr17305 proof

Thanks @zuk3975

File: src/Core/Domain/Supplier/QueryResult/EditableSupplier.php [hunk 1]
```diff
@@ -172,142 +180,151 @@ public function __construct(
...
      */
-    public function getMobilePhone()
+    public function getMobilePhone(): string
     {
         return $this->mobilePhone;
     }
 
     /**
-     * @return array
+     * @return array|null ?array
```
Comment (rokaszygmantas) [inline src/Core/Domain/Supplier/QueryResult/EditableSupplier.php]: strange docblock

**Review by rokaszygmantas — CHANGES_REQUESTED** [review]

File: src/Core/Domain/Supplier/QueryResult/EditableSupplier.php (hunk 1 above)
Comment (zuk3975) [reply src/Core/Domain/Supplier/QueryResult/EditableSupplier.php]: done

**Review by rokaszygmantas — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Pull request merged**